### PR TITLE
Introduce `ComputerListener#onIdle`

### PR DESCRIPTION
**Pull request by Vlatombe**
Introduces a new computer listener fired whenever a `Computer` becomes idle.

We have multiple internal CloudBees usage for this. In Jenkins OSS, one could easily implement a listener to make the retention strategy check fire more quickly than the default cron (1 minute) once an agent becomes idle (when using `Demand` retention strategy for example)

See [JENKINS-XXXXX](https://issues.jenkins.io/browse/JENKINS-XXXXX).

### Proposed changelog entries

- Developer: Add a new listener when a computer becomes idle.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/java/hudson/model/Computer.java [hunk 1]
```diff
@@ -1102,6 +1102,8 @@ protected void removeExecutor(final Executor e) {
                     if (ciBase != null) { // TODO confirm safe to assume non-null and use getInstance()
                         ciBase.removeComputer(Computer.this);
                     }
+                } else if (isIdle()) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java:1105]: Maybe this should be moved out of the `synchronized` block? Or the notification if not the `isIdle` check. Not clear if there is a possible lock ordering problem here. https://github.com/jenkinsci/jenkins/blob/5df30de4358ca93db17eb1d80a6b3a6f1c92ffad/core/src/main/java/hudson/model/Executor.java#L621-L622

**Review by jglick — APPROVED** [review]

File: core/src/main/java/hudson/model/Computer.java (hunk 1 above)
Comment (Vlatombe) [reply core/src/main/java/hudson/model/Computer.java:1105]: Lock ordering is the same as for isAlive() check, and I could look up either individual executor locks, or Queue > Executor lock. Computer > Executor does seem a logical order.
On the other hand, it makes sense to execute the listener outside the lock.

File: core/src/main/java/hudson/model/Computer.java [hunk 2]
```diff
@@ -1103,7 +1105,7 @@ protected void removeExecutor(final Executor e) {
                         ciBase.removeComputer(Computer.this);
                     }
                 } else if (isIdle()) {
-                    Listeners.notify(ComputerListener.class, false, l -> l.onIdle(this));
+                    threadPoolForRemoting.submit(() -> Listeners.notify(ComputerListener.class, false, l -> l.onIdle(this)));
```
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java:1108]: Or just set a flag checked outside the `synchronized` block and run `notify` synchronously then.

Comment (Vlatombe) [reply core/src/main/java/hudson/model/Computer.java:1108]: ok

File: core/src/main/java/hudson/model/Computer.java [hunk 3]
```diff
@@ -1093,6 +1095,7 @@ public final long getDemandStartMilliseconds() {
      */
     protected void removeExecutor(final Executor e) {
         final Runnable task = () -> {
```
Comment (jglick) [inline core/src/main/java/hudson/model/Computer.java:1097]: Hmm, now I see that this whole `task` is also inside the `Queue` lock too. Would be safest to move notification outside there, but then you have to make sure notification happens after the `remove` calls, so it is looking easier to going back to asynch notification. Sorry.

**Review by jglick — APPROVED** [review]

**Comment by Vlatombe** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**